app/lib/pages/homework/homework_dialog.dart

```diff
@@ -30,46 +33,96 @@ import 'package:time/time.dart';
   const HomeworkDialog({
     Key? key,
-    this.homework,
-    required this.homeworkDialogApi,
-    required this.nextLessonCalculator,
+    required this.id,
+    @visibleForTesting this.homeworkDialogApi,
+    @visibleForTesting this.nextLessonCalculator,
   }) : super(key: key);
 
   static const tag = "homework-dialog";
 
-  final HomeworkDto? homework;
-  final HomeworkDialogApi homeworkDialogApi;
-  final NextLessonCalculator nextLessonCalculator;
+  final HomeworkId? id;
+  final HomeworkDialogApi? homeworkDialogApi;
+  final NextLessonCalculator? nextLessonCalculator;
 
   @override
   State createState() => _HomeworkDialogState();
 }
 
 class _HomeworkDialogState extends State<HomeworkDialog> {
   late HomeworkDialogBloc bloc;
+  late Future<HomeworkDto?> homework;
 
   @override
   void initState() {
-    final markdownAnalytics = BlocProvider.of<MarkdownAnalytics>(context);
-    final analytics = BlocProvider.of<SharezoneContext>(context).analytics;
-    bloc = HomeworkDialogBloc(
-      widget.homeworkDialogApi,
-      widget.nextLessonCalculator,
-      markdownAnalytics,
-      homework: widget.homework,
-      analytics,
-    );
     super.initState();
+    final markdownAnalytics = BlocProvider.of<MarkdownAnalytics>(context);
+    final szContext = BlocProvider.of<SharezoneContext>(context);
+    final analytics = szContext.analytics;
+
+    late NextLessonCalculator nextLessonCalculator;
+    if (widget.nextLessonCalculator != null) {
+      nextLessonCalculator = widget.nextLessonCalculator!;
+    } else {
+      final holidayManager =
+          BlocProvider.of<HolidayBloc>(context).holidayManager;
+      nextLessonCalculator = NextLessonCalculator(
+          timetableGateway: szContext.api.timetable,
+          userGateway: szContext.api.user,
+          holidayManager: holidayManager);
+    }
+    final homeworkDialogApi =
+        widget.homeworkDialogApi ?? HomeworkDialogApi(szContext.api);
+
+    if (widget.id != null) {
+      homework = szContext.api.homework
+          .singleHomework(widget.id!.id, source: Source.cache)
+          .then((value) {
+        bloc = HomeworkDialogBloc(
+          homework: value,
+          homeworkDialogApi,
+          nextLessonCalculator,
+          markdownAnalytics,
+          analytics,
+        );
+        return value;
+      });
+    } else {
+      homework = Future.value(null);
+      bloc = HomeworkDialogBloc(
+        homework: null,
+        homeworkDialogApi,
+        nextLessonCalculator,
+        markdownAnalytics,
+        analytics,
+      );
+    }
   }
 
   @override
   Widget build(BuildContext context) {
-    return BlocProvider(
-      bloc: bloc,
-      child: __HomeworkDialog(
-        homework: widget.homework,
-        bloc: bloc,
-      ),
+    return FutureBuilder(
+      future: homework,
+      builder: (context, snapshot) {
+        if (snapshot.hasError ||
+            snapshot.connectionState == ConnectionState.waiting) {
```

Comment:
Is that correct that we display the error message when the state is loading?

Reply:
No this just because both the error and the loading state use mostly the same widgets. There is another conditional below so that only the error text is shown if there is an actual error.